Dockerfile

```diff
@@ -0,0 +1,73 @@
+########################################
+# Stage 1: Build the pathfinder binary #
+########################################
+FROM rust:1.59-alpine AS rust-builder
+
+RUN apk add --no-cache musl-dev gcc openssl-dev
+
+WORKDIR /usr/src/pathfinder
+
+# Build only the dependencies first. This utilizes
+# container layer caching for Rust builds
+RUN mkdir crates
+RUN cargo new --lib crates/pedersen
+# Correct: --lib. We'll handle the binary later.
+RUN cargo new --lib crates/pathfinder
+COPY Cargo.toml Cargo.toml
+
+COPY crates/pathfinder/Cargo.toml crates/pathfinder/Cargo.toml
+COPY crates/pedersen/Cargo.toml crates/pedersen/Cargo.toml
+COPY crates/pedersen/benches crates/pedersen/benches
+
+RUN RUSTFLAGS='-L/usr/lib -Ctarget-feature=-crt-static' cargo build --release --target=x86_64-unknown-linux-musl
+
+
+# Compile the actual libraries and binary now
+COPY . .
+
+# Mark these for re-compilation 
+RUN touch crates/pathfinder/src/lib.rs
+RUN touch crates/pedersen/src/lib.rs
+
+RUN RUSTFLAGS='-L/usr/lib -Ctarget-feature=-crt-static' cargo build --release --target=x86_64-unknown-linux-musl --bin pathfinder
+
+#######################################
+# Stage 2: Build the Python libraries #
+#######################################
+FROM python:3.8-alpine AS python-builder
+
+RUN apk add --no-cache gcc musl-dev gmp-dev g++
+
+WORKDIR /usr/share/pathfinder
+COPY py py
+RUN python3 -m pip install --upgrade pip
+RUN python3 -m pip install -r py/requirements-dev.txt
+
+# This reduces the size of the python libs by about 50%
+ENV PY_PATH=/usr/local/lib/python3.8/
+RUN find ${PY_PATH} -type d -a -name test -exec rm -rf '{}' +
+RUN find ${PY_PATH} -type d -a -name tests  -exec rm -rf '{}' +
+RUN find ${PY_PATH} -type f -a -name '*.pyc' -exec rm -rf '{}' +
+RUN find ${PY_PATH} -type f -a -name '*.pyo' -exec rm -rf '{}' +
+
+#######################
+# Final Stage: Runner #
+#######################
+FROM python:3.8-alpine AS runner
+
+RUN apk add --no-cache tini
+
+COPY --from=rust-builder ["/usr/lib/libstdc++.so.6", "/usr/lib/libgcc_s.so.1", "/usr/lib/libgmp.so.10", "/usr/lib/" ]
+COPY --from=rust-builder /usr/src/pathfinder/target/x86_64-unknown-linux-musl/release/pathfinder /usr/local/bin/pathfinder
+COPY --from=python-builder /usr/local/lib/python3.8/ /usr/local/lib/python3.8/
+
+# Create directory and volume for persistent data
+RUN mkdir -p /usr/share/pathfinder/data
```

Comment:
Hmm, shouldn't this be something like `/var/lib/pathfinder`?

According to the [Linux Filesystem Hierarchy Standard](https://refspecs.linuxfoundation.org/FHS_3.0/fhs/ch04s11.html):

> The /usr/share hierarchy is for all read-only architecture independent data files. [[30]](https://refspecs.linuxfoundation.org/FHS_3.0/fhs/ch04s11.html#ftn.idm236091847456)

And for `/var/lib` ([FHS reference](https://refspecs.linuxfoundation.org/FHS_3.0/fhs/ch05s08.html)):

> This hierarchy holds state information pertaining to an application or the system. State information is data that programs modify while they run, and that pertains to one specific host. Users must never need to modify files in /var/lib to configure a package's operation, and the specific file hierarchy used to store the data must not be exposed to regular users.

Reply:
> This hierarchy holds state information pertaining to an application or the system. State information is data that programs modify while they run, and that pertains to one specific host. **Users must never need to modify files in /var/lib to configure a package's operation**, and the specific file hierarchy used to store the data must not be exposed to regular users.

There should be 2 volumes: one for the configuration in `/usr/share` and one for application data in `/var/lib`.

Reply:
@fracek do you want to make a PR for those volumes? Also note I just created #190 if you want to give any feedback there